src/internal/connector/graph/service.go

```diff
@@ -297,25 +357,23 @@ func (handler *LoggingMiddleware) Intercept(
 	return resp, err
 }
 
-// Run a function with retries
-func RunWithRetry(run func() error) error {
-	var err error
-
-	for i := 0; i < numberOfRetries; i++ {
-		err = run()
-		if err == nil {
-			return nil
-		}
+// Intercept implements the interface and evaluates whether to retry a failed request.
+func (middleware RetryHandler) Intercept(
+	pipeline khttp.Pipeline,
+	middlewareIndex int,
+	req *http.Request,
+) (*http.Response, error) {
+	ctx := req.Context()
 
-		// only retry on timeouts and 500-internal-errors.
-		if !(IsErrTimeout(err) || IsInternalServerError(err)) {
-			break
-		}
+	response, err := pipeline.Next(req, middlewareIndex)
+	if err != nil && !IsErrTimeout(err) {
+		return response, err
+	}
 
-		if i < numberOfRetries {
-			time.Sleep(time.Duration(3*(i+2)) * time.Second)
-		}
+	reqOption, ok := req.Context().Value(retryKeyValue).(RetryHandlerOptions)
+	if !ok {
+		reqOption = middleware.options
```

Comment:
I'm still not sure this is a helpful pattern for us, when these configurations are already accessible through the RetryHandler itself.  It implies that configuration is expected to change on a per-call basis, and I'm not sure we have reason to support that at this time.  Is there a situation where you expect us to need this in the immediate future?